wandb/stats.py

```diff
@@ -40,4 +45,65 @@ def summary(self):
             'total_files': len(self._files),
             'uploaded_bytes': sum(f.uploaded for f in self._files.values()),
             'total_bytes': sum(f.size for f in self._files.values())
-        }
\ No newline at end of file
+        }
+
+
+class SystemStats(object):
+    def __init__(self, run):
+        try:
+            self.gpu_count = nvmlDeviceGetCount()
+        except NVMLError as err:
+            self.gpu_count = 0
+        self.run = run
+        self.sampler = {}
+        self.samples = 0
+        self._thread = threading.Thread(target=self._thread_body)
+        self._thread.daemon = True
+        self._thread.start()
+
+    def _thread_body(self):
+        while True:
+            stats = self.stats()
+            for stat, value in stats.items():
+                if isinstance(value, Number):
+                    self.sampler[stat] = self.sampler.get(stat, [])
+                    self.sampler[stat].append(value)
+            self.samples += 1
+            if self.samples >= 15:
+                self.flush()
+            time.sleep(2)
+
+    def flush(self):
+        stats = self.stats()
+        for stat, value in stats.items():
+            # TODO: a bit hacky, we assume all numbers should be averaged.  If you want
+            # max for a stat, you must put it in a sub key, like ["network"]["sent"]
+            if isinstance(value, Number):
+                samples = self.sampler.get(stat, [stats[stat]])
+                stats[stat] = round(reduce(lambda x, y: x + y,
```

Comment:
This averaging probably introduces a lot of bias in the measurement. If the sample period has high variance (which it may within a single machine and certainly probably does across machines). We can fix a bit by trying to make sure the samples are actually taken uniformly over a 30 second period. To do that you'd mark the time that you took the last sample, and change the sleep to be variable (or use some other timer like you mention). We can fix later but I'd at least file an issue if you want to run with what we have.